src/CommunityToolkit.Maui/Views/Popup/Popup.shared.cs

```diff
@@ -55,7 +65,8 @@ public Popup()
 	{
 		platformConfigurationRegistry = new Lazy<PlatformConfigurationRegistry<Popup>>(() => new(this));
 
-		VerticalOptions = HorizontalOptions = LayoutAlignment.Center;
```

Comment:
Any particular reason to remove this line?

Reply:
@pictos , Thank you for your review.

I removed that code because LayoutAlignment.Center is set as the default value for VerticalOptionsProperty, HorizontalOptionsProperty. Also, if you explicitly specify VerticalOptions and HorizontalOptions before 
applying the style, the style won't be reflected, so I removed that code.

Reply:
I thought it would be overridden if the value were added after the style... Interesting, and how the popup is shown now? Those values were there to make sure the popup renderers at the center of the screen

Reply:
@pictos , I will answer your question.
If you have the code below, the behavior will be the same as when you explicitly specify the value of the Popup property in XAML, and the value specified explicitly will take precedence over the value specified in the style. 

    VerticalOptions = HorizontalOptions = LayoutAlignment.Center;

I verified it again, and if the above code exists, even if HorizontalOptions and VerticalOptions are specified on the style, the Popup will be displayed in the center of the screen regardless of their values.

Although it is unrelated to this case, as a result of verification, we will modify the targetType passed to the MergedStyle constructor as follows.

    mergedStyle = new MergedStyle(GetType(), this);

The reason is that when applying a global style to a class that inherits from Popup, the target style was not found.
The above corrections will be uploaded when the sample code is added.

Sorry for the late reply as I was dealing with another matter at work.

Reply:
When testing with styles were you making use of the `ApplyToDerivedTypes` property on `Style`?

Reply:
@bijington, Since the ApplyToDerivedTypesStyle property was not used during testing, I will specify it and verify it.
Thank you for your comment.

Reply:
I fixed the MergedStyle argument and uploaded it along with the sample page.

Among the things I tested, dynamic style inheritance was the only one that didn't work.
If you specify the BaseResourceKey property of Style, the style will not be applied.
Dynamic style inheritance is not available for Popups, but it is available for the Popup's content.

At least the following will work.

・Implicit Style
・Explicit Style
・Dynamic Style
・ApplyToDerivedTypes
・Style Inheritance
・Style Class

When creating a new Popup, it must be derived from the Popup class, so you can also 
specify the ApplyToDerivedTypes property of Style. 

>We don't have this luxury with Popups because a user can't just create a new instance of Popup, it has to be a derived type.
@bijington , Sorry, I don't understand the deeper meaning of the above.

Reply:
When debugging the dynamic style inheritance pattern, an exception occurred at the following location.

[src\CommunityToolkit.Maui\Views\Popup\Popup.shared.cs]

    public static readonly BindableProperty StyleProperty = BindableProperty.Create(nameof(Style), typeof(Style), typeof(Popup), default(Style), propertyChanged: (bindable, oldValue, newValue) => ((Popup)bindable).mergedStyle.Style = (Style)newValue);

The exception that occurred is as follows.

{System.NullReferenceException: Object reference not set to an instance of an object.
   at Microsoft.Maui.Controls.Style.OnBasedOnResourceChanged(BindableObject bindable, Object oldValue, Object newValue) in D:\a\_work\1\s\src\Controls\src\Core\Style.cs:line 165
   at Microsoft.Maui.Controls.BindableObject.SetValueActual(BindableProperty property, BindablePropertyContext context, Object value, Boolean currentlyApplying, SetValueFlags attributes, Boolean silent) in D:\a\_work\1\s\src\Controls\src\Core\BindableObject.cs:line 535
   at Microsoft.Maui.Controls.BindableObject.SetValueCore(BindableProperty property, Object value, SetValueFlags attributes, SetValuePrivateFlags privateAttributes) in D:\a\_work\1\s\src\Controls\src\Core\BindableObject.cs:line 469
   at Microsoft.Maui.Controls.BindableObject.SetValueCore(BindableProperty property, Object value, SetValueFlags attributes) in D:\a\_work\1\s\src\Controls\src\Core\BindableObject.cs:line 401
   at Microsoft.Maui.Controls.Element.OnResourceChanged(BindableProperty property, Object value) in D:\a\_work\1\s\src\Controls\src\Core\Element.cs:line 634
   at Microsoft.Maui.Controls.Element.OnSetDynamicResource(BindableProperty property, String key) in D:\a\_work\1\s\src\Controls\src\Core\Element.cs:line 507
   at Microsoft.Maui.Controls.BindableObject.SetDynamicResource(BindableProperty property, String key, Boolean fromStyle) in D:\a\_work\1\s\src\Controls\src\Core\BindableObject.cs:line 361
   at Microsoft.Maui.Controls.BindableObject.SetDynamicResource(BindableProperty property, String key) in D:\a\_work\1\s\src\Controls\src\Core\BindableObject.cs:line 342
   at Microsoft.Maui.Controls.Style.Microsoft.Maui.Controls.IStyle.Apply(BindableObject bindable) in D:\a\_work\1\s\src\Controls\src\Core\Style.cs:line 104
   at CommunityToolkit.Maui.Views.MergedStyle2.SetStyle(IStyle implicitStyle, IList`1 classStyles, IStyle style) in C:\Users\tester\Documents\GitHub8\Maui\src\CommunityToolkit.Maui\Views\Popup\Popup.shared.cs:line 704
   at CommunityToolkit.Maui.Views.MergedStyle2.set_Style(IStyle value) in C:\Users\tester\Documents\GitHub8\Maui\src\CommunityToolkit.Maui\Views\Popup\Popup.shared.cs:line 496
   at CommunityToolkit.Maui.Views.Popup.<>c.<.cctor>b__87_0(BindableObject bindable, Object oldValue, Object newValue) in C:\Users\tester\Documents\GitHub8\Maui\src\CommunityToolkit.Maui\Views\Popup\Popup.shared.cs:line 53}

The source is .NET MAUI version 7.0.92.

If you catch the exception and apply the Style, it will work as intended, but that is an unwanted modification.

[src\CommunityToolkit.Maui\Views\Popup\Popup.shared.cs]

    public static readonly BindableProperty StyleProperty = BindableProperty.Create(nameof(Style), typeof(Style), typeof(Popup), default(Style), propertyChanged: (bindable, oldValue, newValue) => {
        try
        {
            ((Popup)bindable).mergedStyle.Style = (Style)newValue;
        }
        catch (NullReferenceException)
        {			
            ((Popup)bindable).mergedStyle.Style.Apply(bindable);
        }
    });

I tested it using the following style.

[App.xaml]

    <Application xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
        xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
        xmlns:popups="clr-namespace:CommunityToolkit.Maui.Sample.Views.Popups"
        xmlns:mct="http://schemas.microsoft.com/dotnet/2022/maui/toolkit"
        x:Class="CommunityToolkit.Maui.Sample.App">
        <Application.Resources>
            <ResourceDictionary>
                <ResourceDictionary.MergedDictionaries>
                    <ResourceDictionary Source="Resources/Styles/Colors.xaml" />
                    <ResourceDictionary Source="Resources/Styles/Styles.xaml" />
                </ResourceDictionary.MergedDictionaries>

                <Style x:Key="stlDynamicBaseStyle" TargetType="{x:Type popups:DynamicStyleInheritancePopup}">
                    <Setter Property="Size" Value="150,150" />
                </Style>
                <Style x:Key="stlDynamicInheritanceStyle1" TargetType="{x:Type popups:DynamicStyleInheritancePopup}" BasedOn="{StaticResource stlDynamicBaseStyle}">
                    <Setter Property="Color" Value="Purple" />
                    <Setter Property="VerticalOptions" Value="End" />
                </Style>
                <Style x:Key="stlDynamicInheritanceStyle2" TargetType="{x:Type popups:DynamicStyleInheritancePopup}" BaseResourceKey="stlDynamicInheritanceStyle1">
                    <Setter Property="Color" Value="Orange" />
                    <Setter Property="HorizontalOptions" Value="End" />
                    <Setter Property="CanBeDismissedByTappingOutsideOfPopup" Value="True" />
                </Style>
            </ResourceDictionary>
    </Application.Resources>
    </Application>

[DynamicStyleInheritancePopup.xaml]

    <mct:Popup xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
        xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
        xmlns:mct="http://schemas.microsoft.com/dotnet/2022/maui/toolkit"
        xmlns:popups="clr-namespace:CommunityToolkit.Maui.Sample.Views.Popups"
        x:Class="CommunityToolkit.Maui.Sample.Views.Popups.DynamicStyleInheritancePopup" Style="{StaticResource stlDynamicInheritanceStyle2}">
        <Grid VerticalOptions="Fill" HorizontalOptions="Fill" BackgroundColor="Transparent">
            <Grid WidthRequest="10" HeightRequest="10" VerticalOptions="Start" HorizontalOptions="Start" BackgroundColor="Blue" />
            <Grid WidthRequest="10" HeightRequest="10" VerticalOptions="Start" HorizontalOptions="End" BackgroundColor="Blue" />
            <Grid WidthRequest="10" HeightRequest="10" VerticalOptions="End" HorizontalOptions="Start" BackgroundColor="Blue" />
            <Grid WidthRequest="10" HeightRequest="10" VerticalOptions="End" HorizontalOptions="End" BackgroundColor="Blue" />
        </Grid>
    </mct:Popup>

The execution results are as follows.

![Screenshot_1694504839](https://github.com/CommunityToolkit/Maui/assets/125236133/c0b0013b-b7a6-4e60-b842-31c726497075)

Although this is an expected result, it may be best to leave dynamic style inheritance non-guaranteed.

.NET MAUI 7.0.92 is already old, so it is unclear whether the issue will be resolved if verified with the latest source code.

Reply:
> I fixed the MergedStyle argument and uploaded it along with the sample page.
> 
> 
> 
> Among the things I tested, dynamic style inheritance was the only one that didn't work.
> 
> If you specify the BaseResourceKey property of Style, the style will not be applied.
> 
> Dynamic style inheritance is not available for Popups, but it is available for the Popup's content.
> 
> 
> 
> At least the following will work.
> 
> 
> 
> ・Implicit Style
> 
> ・Explicit Style
> 
> ・Dynamic Style
> 
> ・ApplyToDerivedTypes
> 
> ・Style Inheritance
> 
> ・Style Class
> 
> 
> 
> When creating a new Popup, it must be derived from the Popup class, so you can also 
> 
> specify the ApplyToDerivedTypes property of Style. 
> 
> 
> 
> >We don't have this luxury with Popups because a user can't just create a new instance of Popup, it has to be a derived type.
> 
> 
> 
> @bijington , Sorry, I don't understand the deeper meaning of the above.

I was trying compare how styles are used for existing controls like `Label`, etc. where a developer can create an instance of Label by adding it into their UI. They also have the ability to subclass Label let's call it `TitleLabel` and define a style for that. This is where `ApplyToDerivedTypes` makes sense. With `Popup` however a developer cannot simply add one to their UI and they have to create a subclass. Sorry I realise I sent you down the route of using ApplyToDerivedTypes but I am just now questioning whether it makes sense to make use of it. Part of me says we should and another part questions what should happen if that property is set to false, as this would result in a style definition that would never be used. Does that make sense?

Reply:
@bijington , Thank you for your reply. I understand what your comment means.

[Case 1]
If the value of the ApplyToDerivedTypes property is True, the style defined with the value of the ApplyToDerivedTypes property of True is applied unless a style that targets the derived type exists.

[Case 2]
If the value of the ApplyToDerivedTypes property is True and a style that targets a derived type is defined, the style that targets the derived type takes precedence.

[Case 3]
If the ApplyToDerivedTypes property is False, no styles are applied to derived types, so the value you explicitly set for the property has meaning.

Case 1 is used when you want to specify the same position, size, color, and closing behavior for all derived types. I think this is the case when you want to specify it in cases where it is a hassle to set Popup properties for each derived type one by one.

Case 2 is used when you want to explicitly apply a style when the same style has been applied to all derived types. For example, you might want to change the position of a Popup even though the size and color are the same.

Case 3 makes sense if you generate the Popup content in the code-behind without deriving the type from Popup. There are two ways to design the content of a Popup, but I think it is a case where the Popup is generated in the code-behind without inheriting the Popup class.

I do not choose Case 3 because I choose to design Popup content using XAML. I think Case 3 is necessary depending on how the screen is designed. This is just my personal opinion, but I don't think there is any problem with having the ApplyToDerivedTypes property.

Reply:
https://github.com/CommunityToolkit/Maui/pull/1351#discussion_r1322602255
@bijington , What do you think about this?
If dynamic style inheritance is used, no styles will be applied and an exception will be raised. 
I think something needs to be done, but what should I do?

Reply:
@bijington , It seems that the problem with Dynamic Style can be solved without any tricks.
That is to inherit IStyleElement to the Popup class. 

[src\Controls\src\Core\Style.cs]

    static void OnBasedOnResourceChanged(BindableObject bindable, object oldValue, object newValue)
    {
        Style style = (bindable as IStyleElement).Style;     // <= here
        if (style == null)
            return;
        if (!style._targets.TryGetValue(bindable, out var objectspecificity))
            return;

        style.UnApplyCore(bindable, (Style)oldValue);
        style.ApplyCore(bindable, (Style)newValue, (SetterSpecificity)objectspecificity);
    }

When I tried to cast the bindable in the above source code to IStyleElement, it became null.
This is because Popup does not inherit IStyleElement.
I will add Dynamic Style to the sample and upload the modified code by the end of today.
Please wait a moment now.

Reply:
@bijington , I just posted the code to add support for Dynami style inheritance.
I made the Popup class inherit IStyleElement. I also added a Dynamic style inheritance pattern to the sample.

Reply:
Looking good!